File: modular_exponential/modularExponential.js
function modularExponential (base, power, mod) {
	/*
	Performs Modular Exponential.
	Time Complexity : O(log (power))
	:param base: Number that is going to be raised.
	:param power: Power to which the number is raised.
	:param mod: Number by which modulo has to be performed.
	:return: Returns (base ^ power) % mod
	*/
	let answer = 1;
	base = base % mod;
	while (power) {
		if (power & 1) {
			answer = (answer * base) % mod;
		}
		power = power >> 1;
		base = (base * base) % mod;
	}
	return answer;
}

function main () {
	let base = 3;
	let power = 5;
	let mod = 61;
	let res = modularExponential(base, power, mod);
	console.log('%d\n', res);
}

main();

File: modular_exponential/modular_exponential.c
#include <stdio.h>

// Time Complexity : O(log (power))
long long modularExponential(long long base, long power, long long mod) {
	long long answer = 1;
	base = base % mod;
	while (power) {
		if (power & 1) {
			answer = (answer * base) % mod;
		}
		power = power >> 1;
		base = (base * base) % mod;
	}
	return answer;
}

int main() {
	long long base = 2;
	long power = 10;
	long long mod = 100000;
	printf("%lld\n", modularExponential(base, power, mod));
	return 0;
}
